packages/flutter/lib/src/widgets/scrollbar.dart

```diff
@@ -1838,7 +2168,110 @@ class _TrackTapGestureRecognizer extends TapGestureRecognizer {
+      return false;
+    }
+    return super.isPointerAllowed(event);
+  }
+
+  bool _hitTestInteractive(GlobalKey customPaintKey, Offset offset, PointerDeviceKind kind) {
+    if (customPaintKey.currentContext == null) {
+      return false;
+    }
+    final CustomPaint customPaint = customPaintKey.currentContext!.widget as CustomPaint;
+    final ScrollbarPainter painter = customPaint.foregroundPainter! as ScrollbarPainter;
+    final Offset localOffset = _getLocalOffset(customPaintKey, offset);
+    return painter.hitTestButtonsInteractive(localOffset, kind, _isLeadingButton);
+  }
+}
+
+// A long press gesture detector that responds to events on the leading scrollbar's button.
+class _LeadingButtonPressGestureRecognizer extends _ButtonPressGestureRecognizer {
+  _LeadingButtonPressGestureRecognizer({
+    double? postAcceptSlopTolerance,
+    Set<PointerDeviceKind>? supportedDevices,
+    required Object debugOwner,
+    required GlobalKey customPaintKey,
+    required Duration pressDuration,
+  }) : super(
+         true,
+         postAcceptSlopTolerance: postAcceptSlopTolerance,
+         supportedDevices: supportedDevices,
+         debugOwner: debugOwner,
+         customPaintKey: customPaintKey,
+         pressDuration: pressDuration,
+       );
+}
+
+// A long press gesture detector that responds to events on the trailing scrollbar's button.
+class _TrailingButtonPressGestureRecognizer extends _ButtonPressGestureRecognizer {
+  _TrailingButtonPressGestureRecognizer({
+    double? postAcceptSlopTolerance,
+    Set<PointerDeviceKind>? supportedDevices,
+    required Object debugOwner,
+    required GlobalKey customPaintKey,
+    required Duration pressDuration,
+  }) : super(
+    false,
+    postAcceptSlopTolerance: postAcceptSlopTolerance,
+    supportedDevices: supportedDevices,
+    debugOwner: debugOwner,
+    customPaintKey: customPaintKey,
+    pressDuration: pressDuration,
+  );
+}
+
 Offset _getLocalOffset(GlobalKey scrollbarPainterKey, Offset position) {
   final RenderBox renderBox = scrollbarPainterKey.currentContext!.findRenderObject()! as RenderBox;
   return renderBox.globalToLocal(position);
 }
+
+/// Base class for Scrollbar leading and trailing buttons shapes.
+///
+abstract class ScrollbarButtonsShape {
+  /// This abstract const constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const ScrollbarButtonsShape();
+
+  /// Return the button size.
+  double getButtonSize() => 18.0;
+
+  /// Paints the buttons shape based on the state passed to it.
+  ///
+  void paint(
+    Canvas canvas,
+    Rect leadingButtonRect,
+    bool isLeadingButtonHovered,
+    bool isLeadingButtonPressed,
+    Rect trailingButtonRect,
+    bool isTrailingButtonHovered,
+    bool isTrailingButtonPressed,
+    Paint thumbPaint,
+    Paint trackPaint,
+    Paint borderPaint,
+    ScrollMetrics scrollMetrics,
+    TextDirection textDirection,
+    ScrollbarOrientation orientation,
```

Comment:
Yeah I think this could be simplified if you can create an object that describes whatever the painter needs to know about the buttons, what the state of each one is, what is each rect, etc.